✨ CBPK Use Unstructured to access Config Owner

juan-lee opened the pull request:
**What this PR does / why we need it**:
In preparation for adding `MachinePool`, update CBPK to access the configuration owner through `Unstructured` instead of `Machine` directly. `Machine`, `MachinePool`, or other are now required to implement the following fields:
- metadata.labels["cluster.x-k8s.io/control-plane"]
- spec.clusterName
- spec.bootstrap.dataSecretName
- status.infrastructureReady

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1799

juan-lee commented:
@CecileRobertMichon @devigned @ritazh FYI

neolit123 commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -125,28 +125,28 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
-	if machine == nil {
-		log.Info("Waiting for Machine Controller to set OwnerRef on the KubeadmConfig")
+	if configOwner == nil {
+		log.Info("Waiting for Owning Controller to set OwnerRef on the KubeadmConfig")
```
is "Owning" the actual name of the controller :thinking: ?

juan-lee replied:
nope, I changed the string to be more specific. PTAL @neolit123

neolit123 commented on util/configowner_test.go:
```diff
@@ -0,0 +1,132 @@
+func TestGetConfigOwnerNotFound(t *testing.T) {
+	scheme := runtime.NewScheme()
+	if err := clusterv1.AddToScheme(scheme); err != nil {
+		t.Fatal("failed to register cluster api objects to scheme")
```
is "cluster api" consistent naming across user facing output, including unit tests?

juan-lee replied:
I changed the string to "cluster-api". I don't have the answer to your question, @detiber might be able to comment.

ncdc replied:
I don't believe this is something where consistency matters, especially since it's a failure message in a unit test

detiber replied:
probably should use "Cluster API", however I'm not worried about test output for this, only user output for released binaries.

detiber commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -125,28 +125,28 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, nil
 	}
-	log = log.WithValues("machine-name", machine.Name)
+	log = log.WithValues("config-owner-name", configOwner.GetName())
```
Would it make sense to pull the type information from the unstructured resource and add it to the logging in addition to the name?
maybe something like
```go
log = log.WithValues("kind", configOwner.Kind, "version", configOwner.GetResourceVersion(), "name", configOwner.GetName())?
```

detiber commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -125,28 +125,28 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	if err != nil {
 		if errors.Cause(err) == util.ErrNoCluster {
-			log.Info("Machine does not belong to a cluster yet, waiting until its part of a cluster")
+			log.Info("Config owner does not belong to a cluster yet, waiting until its part of a cluster")
```
Might be better to use the kind from the resource here instead: `configOwner.GetKind()`

detiber commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -255,10 +255,15 @@ func (r *KubeadmConfigReconciler) handleClusterNotInitialized(ctx context.Contex
 
+	machine := &clusterv1.Machine{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(scope.ConfigOwner.Object, machine); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "cannot convert object to %s", scope.ConfigOwner.GetKind())
```
Wouldn't this error message return "cannot convert object to MachinePool" if a MachinePool resource makes it to here?

juan-lee replied:
good catch, thanks :)

detiber commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -255,10 +255,15 @@ func (r *KubeadmConfigReconciler) handleClusterNotInitialized(ctx context.Contex
 	}
 
+	machine := &clusterv1.Machine{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(scope.ConfigOwner.Object, machine); err != nil {
```
Would it be good to do a check that `scope.ConfigOwner.GetKind() == "Machine"` here first?

juan-lee replied:
Yes, for now I'll restrict control plane related operations to Machine since we don't plan on supporting MachinePool control planes in the immediate future.

juan-lee commented:
/assign @vincepri

detiber commented:
/lgtm

vincepri commented on util/configowner.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+// IsControlPlane checks if an unstructured object is a control plane node.
+func (co ConfigOwner) IsControlPlane() bool {
```
Should this be more like `IsControlPlaneMachine` instead? We'd include the current check for labels, and the check that the owner's `Kind` is a Machine.

vincepri commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -255,10 +255,19 @@ func (r *KubeadmConfigReconciler) handleClusterNotInitialized(ctx context.Contex
 		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 	}
 
+	if scope.ConfigOwner.GetKind() != "Machine" {
```
If we switch to `IsControlPlaneMachine`, this check can be removed

vincepri commented on util/configowner.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+// InfrastructureReady extracts infrastructure status from the config owner.
+func (co ConfigOwner) InfrastructureReady() bool {
```
```suggestion
func (co ConfigOwner) IsInfrastructureReady() bool {
```

for consistency

vincepri commented on util/configowner.go:
```diff
@@ -0,0 +1,90 @@
+	return nil, nil
+}
+
+// GetUnstructuredByRef finds and return a Unstructured object using the specified params.
```
```suggestion
// GetUnstructuredByRef finds and returns the owner by looking at the object reference.
```

vincepri commented on util/configowner.go:
```diff
@@ -0,0 +1,90 @@
+/*
```
This file should probably be moved under `bootstrap/util` instead, given that it seems very specific to what the Kubeadm bootstrapper needs.

wdyt?

vincepri commented:
This change and the necessary modifications need to be documented for other bootstrap providers in the v1a2->v1a3 document

juan-lee commented:
> This change and the necessary modifications need to be documented for other bootstrap providers in the v1a2->v1a3 document

added a section in the v1a2->v1a3 doc and addressed feedback @vincepri @detiber PTAL

detiber commented on bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:
```diff
@@ -409,6 +415,10 @@ func (r *KubeadmConfigReconciler) joinWorker(ctx context.Context, scope *Scope)
 }
 
 func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *Scope) (ctrl.Result, error) {
+	if scope.ConfigOwner.GetKind() != "Machine" {
```
Use `scope.ConfigOwner.IsControlPlaneMachine()` here?

juan-lee commented:
/assign @detiber

detiber commented:
/lgtm
ping @vincepri for review for approval.

vincepri commented:
/approve
